templates/commands/render/action_include.go

```diff
@@ -35,93 +34,133 @@ func actionInclude(ctx context.Context, inc *spec.Include, sp *stepParams) error
-		// len(asPaths) is either == 0 or == len(incPaths).
-		as := p.Val
-		if len(asPaths) > 0 {
-			as = asPaths[i].Val
-		}
+	incPaths, err := processPaths(inc.Paths, sp.scope)
+	if err != nil {
+		return err
+	}
 
-		// By default, we copy from the template directory. We also support
-		// grabbing files from the destination directory, so we can modify files
-		// that already exist in the destination.
-		fromDir := sp.templateDir
-		if inc.From.Val == "destination" {
-			fromDir = sp.flags.Dest
-		}
-		absSrc := filepath.Join(fromDir, p.Val)
-		absDst := filepath.Join(sp.scratchDir, as)
-
-		skipNow := maps.Clone(skip)
-		if absSrc == sp.templateDir {
-			// If we're copying the template root directory, automatically skip
-			// 1. spec.yaml file, because it's very unlikely that the user actually
-			// wants the spec file in the template output.
-			// 2. testdata/golden directory, this is reserved for golden test usage.
-			skipNow["spec.yaml"] = struct{}{}
-			skipNow[filepath.Join("testdata", "golden")] = struct{}{}
+	for i, p := range incPaths {
+		matchedPaths, err := processGlobs(ctx, []model.String{p}, fromDir)
+		if err != nil {
+			return err
 		}
 
-		if _, err := sp.fs.Stat(absSrc); err != nil {
-			if common.IsStatNotExistErr(err) {
-				return p.Pos.Errorf("include path doesn't exist: %q", p.Val)
+		for _, matchedPath := range matchedPaths {
+			absSrc := matchedPath.Val
+			relSrc, err := filepath.Rel(fromDir, absSrc)
+			if err != nil {
+				return fmt.Errorf("internal error making relative glob matched path: %w", err)
 			}
-			return fmt.Errorf("Stat(): %w", err)
-		}
-
-		params := &common.CopyParams{
-			DryRun:  false,
-			DstRoot: absDst,
-			RFS:     sp.fs,
-			SrcRoot: absSrc,
-			Visitor: func(relToAbsSrc string, de fs.DirEntry) (common.CopyHint, error) {
-				if _, ok := skipNow[relToAbsSrc]; ok {
-					return common.CopyHint{
-						Skip: true,
-					}, nil
-				}
-
-				abs := filepath.Join(absSrc, relToAbsSrc)
-				relToFromDir, err := filepath.Rel(fromDir, abs)
-				if err != nil {
-					return common.CopyHint{}, fmt.Errorf("filepath.Rel(%s,%s)=%w", fromDir, abs, err)
-				}
-				if !de.IsDir() && inc.From.Val == "destination" {
-					sp.includedFromDest = append(sp.includedFromDest, relToFromDir)
-				}
+			relDst := getRelDst(asPaths, i, p, matchedPaths, relSrc, absSrc, fromDir)
+			absDst := filepath.Join(sp.scratchDir, relDst)
 
-				return common.CopyHint{
-					// Allow later includes to replace earlier includes in the
-					// scratch directory. This doesn't affect whether files in
-					// the final *destination* directory will be overwritten;
-					// that comes later.
-					Overwrite: true,
-				}, nil
-			},
-		}
-		if err := common.CopyRecursive(ctx, p.Pos, params); err != nil {
-			return p.Pos.Errorf("copying failed: %w", err)
+			if err := copyToDst(ctx, sp, skip, matchedPath.Pos, absDst, absSrc, relSrc, inc.From.Val, fromDir); err != nil {
+				return err
+			}
 		}
 	}
 	return nil
```

Comment:
nit
```suggestion
			return copyToDst(ctx, sp, skip, matchedPath.Pos, absDst,
			  absSrc, relSrc, inc.From.Val, fromDir)
```

Reply:
oh yeah, nice one

Reply:
wait, I think this actually fails because when err == nil, it stops the matchedPath loop.